core/client/src/main/java/alluxio/client/file/UnderFileSystemFileInStream.java

```diff
@@ -164,7 +164,10 @@ private ByteBuffer allocateBuffer() {
   private void bufferedRead(int len) throws IOException {
     mBuffer.clear();
     int bytesRead = directRead(mBuffer.array(), 0, len);
-    mBuffer.limit(bytesRead);
+    if (bytesRead >= 0) {
+      mBuffer.limit(bytesRead);
+      mBuffer.position(0);
```

Comment:
Since we are updating the backing array directly, we don't have any guarantees that `flip` will work correctly.